apply_dimension.json

```diff
@@ -37,7 +38,14 @@
             "required": true
         },
         "dimension": {
-            "description": "The name of the dimension to apply the process on. By default, applies the process on all pixel values (as ``apply()`` does).",
+            "description": "The name of the dimension to apply the process on. Fails with a `DimensionNotAvailable` error if the specified dimension does not exist.",
+            "schema": {
+                "type": "string"
+            },
+            "required": true
+        },
+        "target_dimension": {
+            "description": "The name of the target dimension or `null` to replace the dimension specified in `dimension` (default).\n\nIf a target dimension is specified, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created. If a dimension with the specified name exists, the dimension is replaced, but keeps the original type.\n\nThe resulting target dimension has as many dimension labels as the array returned by the process has values. For example, a process returning an array with four elements results in a dimension with four labels.",
```

Comment:
proposal to simplify things a bit

> The new name of the dimension that replaces the original dimension, or `null` (the default) to reuse the original name and type of the dimension. 
>
> The resulting target dimension has as many dimension labels ...

Reply:
Hmm, yes, there's room for improvement. 
First, the first sentence is better capturing my intention as follows:
> The name of the target dimension or `null` to replace the dimension labels and pixel values in the original dimension (default)

I didn't really intend a drop&add (= replace) of the dimension. If that's desired, specify a target_dimension.

What I'm also thinking about is sorting, for example. It's just the values that change, but also reference system etc should stay intact. What I'm no wondering about is: Should the dimension labels be preserved when sorting? What about preserving the labels when computing the cumsum, for example? This is not really captured any longer with the "enumerated" labels. Hmm...

Reply:
ok, I indeed was wondering about the intended precedence of that sentence "name or null to replace ...":
- (name or null) to replace
- name or (null to replace)

I assumed the first, but you meant the second

Reply:
If I understand correctly, there are quite some aspects you want to control through the `target_dimension` argument:
- dimension name (reuse original or rename)
- dimension type (reuse original or "other")
- dimension labels (reuse originals, new integer range, something else?)

I guess it's a bit much to try to encode this in a single string/null type argument.
Maybe `target_dimension` should be an object with dedicated (but optional) fields, to control each of these aspects?

Reply:
Thanks for the feedback. I tried to make it more clear:

> The name of the target dimension or `null` (default).
> 
> 1. The default behavior is to replace the dimension labels and pixel values in the original dimension.
> 2. If a target dimension is specified, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created. If a dimension with the specified name exists, the dimension is replaced, but keeps the original type.
> 
> Afterwards, the dimension has as many dimension labels as the array returned by the process has values. For example, a process returning an array with four elements results in a dimension with four labels.

Reply:
> I guess it's a bit much to try to encode this in a single string/null type argument.
> Maybe `target_dimension` should be an object with dedicated (but optional) fields, to control each of these aspects?

No, if you want to control these settings, use add_dimension beforehand:
> If a dimension with the specified name exists, the dimension is replaced, but keeps the original settings.

I clarified the target_dimension description again and hope it covers all cases now (see also https://github.com/Open-EO/openeo-processes/pull/111#issuecomment-567473532):
> The name of the target dimension or `null` (default).
>
> 1. The default behavior is to replace the pixel values (and dimension labels if the number of pixel values and dimension labels doesn't match) in the original dimension.
> 2. If a target dimension is specified, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created. If a dimension with the specified name exists, the dimension is replaced, but keeps the original name, type, reference system (and dimension labels if the number of values matches the number of dimension labels).
> 
>Afterwards, the dimension has as many dimension labels as the array returned by the process has values. For example, a process returning an array with four elements results in a dimension with four labels.

Does that make more sense?

Reply:
> If a target dimension is specified, a new dimension ... is created

Maybe I'm reading this wrong, but this allows the interpretation that the old dimension/data is preserved (the new dimension is just added), which practically and mathematically makes no sense in the general case. 

> If a dimension with the specified name exists, the dimension is replaced ....

I don't get this. How can you replace another dimension than the one specified by `dimension` argument? For example you have cube with dims "foo" and "bar", you do apply_dimension on "foo", and you use "bar" as target_dimension, that doesn't make sense to me.

Reply:
You are right, the source dimension would have to be removed. I'll clarify that...

Reply:
Tried to update the description for target_dimension:
> The name of the target dimension or `null` (the default) to use the source dimension specified in the parameter `dimension`.
>
> * The default is that the target dimension is the source dimension so that the pixel values are replaced in the source dimension. The original name, type and reference system are preserved.
> * By specifying a target dimension, the source dimension is removed and the target dimension is populated with the computed pixel values.
>     * If the target dimension *does exist*, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created.
>     * If the target dimension *does not exist*, the pixel values are replaced in the target dimension. The original name, type and reference system are preserved.
>
> In general, the target dimension will have as many dimension labels as values are available after applying the process. The dimension labels are preserved when the number of pixel values in the source dimension is equal to the number of values computed by the process. Otherwise, the dimension labels will be incrementing integers starting from zero.

Does it make more sense now?

Reply:
Yes, that's better. 

But overall it seems like a very long explanation for what is basically a name change. 
For example, isn't it possible to merge the part "if the target dim does exist ..." into the part about the default  behaviour? The case where the target dimension does exist but is different from the source dim should be illegal anyway, right?

Reply:
You are really pushing me into short and crisp writing, @soxofaan ;-)

Do you like commit https://github.com/Open-EO/openeo-processes/pull/111/commits/806bb7dde8bcd8df606d5c0f7b6f3881cca48752 better? I hope I still cover all cases good enough...